soa-derive-internal/src/input.rs

```diff
@@ -230,4 +250,38 @@ impl Input {
     pub fn ptr_mut_name(&self) -> Ident {
         Ident::new(&format!("{}PtrMut", self.name), Span::call_site())
     }
-}
+    pub fn fold_fields(&self, 
+        gen_token_stream: impl Fn(&Ident, &Type, bool) -> proc_macro2::TokenStream, 
+        f: impl Fn(&mut proc_macro2::TokenStream, proc_macro2::TokenStream)
+    ) -> proc_macro2::TokenStream
+    {
+        let mut seq = None;
+        self.fields.iter().zip(self.field_is_nested.iter())
+            .for_each(|(field, is_nested)| {
+                let field_ident = field.ident.as_ref().unwrap();
+                let field_type = &field.ty;
+
+                let next = gen_token_stream(field_ident, field_type, *is_nested);
+                match seq.as_mut() {
+                    None => {
+                        seq = Some(next);
+                    }
+                    Some(seq) => {
+                        f(seq, next);
+                    }
+                }
+            });
+        seq.unwrap()
+    }
+    pub fn fields_seq(&self, 
```

Comment:
Same as above, what is this function doing?

Reply:
Replaced by a combination of `self.iter_fields().map(|(ident, ty, is_nested)| /* generate the tokenstreams */).concat()`, which may be more intuitive